src/node.rs

```diff
@@ -1224,42 +1252,112 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EthNamespaceT for
         let inner = Arc::clone(&self.inner);
 
         Box::pin(async move {
-            let reader = match inner.read() {
-                Ok(r) => r,
-                Err(_) => return Err(into_jsrpc_error(Web3Error::InternalError)),
+            let maybe_block = {
+                let reader = match inner.read() {
+                    Ok(r) => r,
+                    Err(_) => return Err(into_jsrpc_error(Web3Error::InternalError)),
+                };
+                match block_number {
+                    zksync_types::api::BlockNumber::Latest
+                    | zksync_types::api::BlockNumber::Pending
+                    | zksync_types::api::BlockNumber::Finalized
+                    | zksync_types::api::BlockNumber::Committed => reader
+                        .block_hashes
+                        .get(&reader.current_miniblock.saturating_sub(1))
```

Comment:
Why subtracting `1` here? Wouldn't the latest block just be `reader.current_miniblock`?

Reply:
Perhaps it's a misunderstanding, but when we produce a block we use the `current_miniblock` value and only _after_ increment it. So, it seems that the block number is incremented to prepare for the next block.

Reply:
That's true. The confusion (for me) is from only batching a single transaction for each batch, so should the `current_miniblock` be empty or should it be the latest block? And if `current_miniblock` is being incremented after the block is produced, then this is correct and we should subtract 1.

I think this will create a slight deviation from before when someone calls `eth_getBlockNumber` and gets `current_miniblock`... to then use that value and request `eth_getBlockByNumber`. Before, I believe we would return the latest (non-empty) block. The scenario that hit this prior was MetaMask confirmation of transactions, so going to validate from your PR and post an update here. If that works as expected, then we should leave the subtraction of 1 as-is.

Reply:
Currently, to fix the e2e tests in `run` mode I'm inserting a "dummy" genesis block with block number `0`. We can perhaps leave the `current_miniblock` = `0` in that case then this subtraction isn't required. Additionally in `fork` mode we are currently incrementing the miniblock number, so that needs to not happen as well. I personally prefer having `current_miniblock` to be the value of the latest block, instead of the hinted `next_miniblock` semantics of it.

Reply:
After some investigation/debugging, it looks like MetaMask is now unable to connect because it first calls `eth_blockNumber` (which returns `1`), and then calls `eth_getBlockByNumber(1)` and is throws an error that the block doesn't exist...

So, as to never reference to a non-existent block when `eth_getBlockByNumber` is passed in `current_miniblock`, there are a few options:

1. We update `eth_blockNumber` to return `current_miniblock.saturating_sub(1)` and remove `.saturating_sub(1)` from the `get_block_by_number` function
2. We update the block creation logic to increment `current_miniblock` before generating the block
3. We create a `Pending` empty block with the block number of `current_miniblock` and update the previous block with the new transaction and set it to `Committed`.

I'm leaning towards Option 1. The problem with Option 3 is figuring out how to determine what `Latest` is, such that it contains useful information. Option 2 sounds like it would just create a mismatch between batch and miniblock counter.

Reply:
> I personally prefer having `current_miniblock` to be the value of the latest block, instead of the hinted `next_miniblock` semantics of it.

I think this would mean Option 2 listed out above where we increment before generating the block/batch. I'm also fine with this.